File: library/src/main/java/com/akexorcist/roundcornerprogressbar/common/AnimatedRoundCornerProgressBar.java
package com.akexorcist.roundcornerprogressbar.common;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.ValueAnimator;
import android.content.Context;
import android.content.res.TypedArray;
import android.os.Build;
import android.os.Parcel;
import android.os.Parcelable;
import android.util.AttributeSet;
import android.widget.LinearLayout;

import androidx.annotation.FloatRange;
import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.customview.view.AbsSavedState;

import com.akexorcist.roundcornerprogressbar.R;

@SuppressWarnings("unused")
public abstract class AnimatedRoundCornerProgressBar extends BaseRoundCornerProgressBar {
    public static final long DEFAULT_DURATION = 500L;

    private boolean isProgressAnimating = false;
    private boolean isSecondaryProgressAnimating = false;
    private float lastProgress;
    private float lastSecondaryProgress;

    private float animationSpeedScale;
    private boolean isAnimationEnabled;

    private ValueAnimator progressAnimator;
    private ValueAnimator secondaryProgressAnimator;

    public AnimatedRoundCornerProgressBar(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public AnimatedRoundCornerProgressBar(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
    public AnimatedRoundCornerProgressBar(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    @Override
    public void setupStyleable(@NonNull Context context, @NonNull AttributeSet attrs) {
        super.setupStyleable(context, attrs);
        TypedArray typedArray = context.obtainStyledAttributes(attrs, R.styleable.AnimatedRoundCornerProgressBar);
        isAnimationEnabled = typedArray.getBoolean(R.styleable.AnimatedRoundCornerProgressBar_rcAnimationEnable, false);
        animationSpeedScale = typedArray.getFloat(R.styleable.AnimatedRoundCornerProgressBar_rcAnimationSpeedScale, 1f);
        typedArray.recycle();
        lastProgress = super.getProgress();
        lastSecondaryProgress = super.getSecondaryProgress();
    }

    @Override
    public float getProgress() {
        if (!isAnimationEnabled && !isProgressAnimating) {
            return super.getProgress();
        } else {
            return lastProgress;
        }
    }

    @Override
    public void setProgress(int progress) {
        setProgress((float) progress);
    }

    @Override
    public void setProgress(float progress) {
        if (progress < 0) {
            progress = 0;
        } else {
            progress = Math.min(progress, max);
        }
        clearProgressAnimation();
        this.lastProgress = progress;
        if (isAnimationEnabled) {
            startProgressAnimation(super.getProgress(), progress);
        } else {
            super.setProgress(progress);
        }
    }

    @Override
    public float getSecondaryProgress() {
        if (!isAnimationEnabled && !isSecondaryProgressAnimating) {
            return super.getSecondaryProgress();
        } else {
            return lastSecondaryProgress;
        }
    }

    @Override
    public void setSecondaryProgress(int progress) {
        setSecondaryProgress((float) progress);
    }

    @Override
    public void setSecondaryProgress(float progress) {
        if (progress < 0) {
            progress = 0;
        } else {
            progress = Math.min(progress, max);
        }
        clearSecondaryProgressAnimation();
        this.lastSecondaryProgress = progress;
        if (isAnimationEnabled) {
            startSecondaryProgressAnimation(super.getSecondaryProgress(), progress);
        } else {
            super.setSecondaryProgress(progress);
        }
    }

    @FloatRange(from = 0.2f, to = 5f)
    public float getAnimationSpeedScale() {
        return animationSpeedScale;
    }

    public void enableAnimation() {
        this.isAnimationEnabled = true;
    }

    public void disableAnimation() {
        this.isAnimationEnabled = false;
    }

    public void setAnimationSpeedScale(@FloatRange(from = 0.2f, to = 5f) float scale) {
        scale = Math.min(scale, 5);
        scale = Math.max(scale, 0.2f);
        this.animationSpeedScale = scale;
    }

    public boolean isProgressAnimating() {
        return isProgressAnimating;
    }

    public boolean isSecondaryProgressAnimating() {
        return isSecondaryProgressAnimating;
    }

    protected void onProgressChangeAnimationUpdate(LinearLayout layout, float current, float to) {
    }

    protected void onProgressChangeAnimationEnd(LinearLayout layout) {
    }

    protected void stopProgressAnimationImmediately() {
        clearProgressAnimation();
        clearSecondaryProgressAnimation();
        if (isAnimationEnabled && isProgressAnimating) {
            disableAnimation();
            if (isProgressAnimating) {
                super.setProgress(lastProgress);
            }
            if (isSecondaryProgressAnimating) {
                super.setSecondaryProgress(lastProgress);
            }
            enableAnimation();
        }
    }

    private long getAnimationDuration(float from, float to, float max, float scale) {
        float diff = Math.abs(from - to);
        return (long) (diff * DEFAULT_DURATION / max * scale);
    }

    private void startProgressAnimation(float from, float to) {
        isProgressAnimating = true;
        if (progressAnimator != null) {
            progressAnimator.removeUpdateListener(progressAnimationUpdateListener);
            progressAnimator.removeListener(progressAnimationAdapterListener);
            progressAnimator.cancel();
            progressAnimator = null;
        }
        progressAnimator = ValueAnimator.ofFloat(from, to);
        progressAnimator.setDuration(getAnimationDuration(from, to, max, animationSpeedScale));
        progressAnimator.addUpdateListener(progressAnimationUpdateListener);
        progressAnimator.addListener(progressAnimationAdapterListener);
        progressAnimator.start();
    }

    private ValueAnimator.AnimatorUpdateListener progressAnimationUpdateListener = new ValueAnimator.AnimatorUpdateListener() {
        @Override
        public void onAnimationUpdate(ValueAnimator animation) {
            onUpdateProgressByAnimation((float) animation.getAnimatedValue());
        }
    };

    private AnimatorListenerAdapter progressAnimationAdapterListener = new AnimatorListenerAdapter() {
        @Override
        public void onAnimationEnd(Animator animation) {
            isProgressAnimating = false;
            onProgressAnimationEnd();
        }

        @Override
        public void onAnimationCancel(Animator animation) {
            isProgressAnimating = false;
        }
    };

    private void onUpdateProgressByAnimation(float progress) {
        super.setProgress(progress);
        onProgressChangeAnimationUpdate(layoutProgress, progress, lastProgress);
    }

    private void onProgressAnimationEnd() {
        onProgressChangeAnimationEnd(layoutProgress);
    }

    private void restoreProgressAnimationState() {
        if (isProgressAnimating) {
            startProgressAnimation(super.getProgress(), lastProgress);
        }
    }

    private void clearProgressAnimation() {
        if (progressAnimator != null && progressAnimator.isRunning()) {
            progressAnimator.cancel();
        }
    }

    private void startSecondaryProgressAnimation(float from, float to) {
        isSecondaryProgressAnimating = true;
        if (secondaryProgressAnimator != null) {
            secondaryProgressAnimator.removeUpdateListener(secondaryProgressAnimationUpdateListener);
            secondaryProgressAnimator.removeListener(secondaryProgressAnimationAdapterListener);
            secondaryProgressAnimator.cancel();
            secondaryProgressAnimator = null;
        }
        secondaryProgressAnimator = ValueAnimator.ofFloat(from, to);
        secondaryProgressAnimator.setDuration(getAnimationDuration(from, to, max, animationSpeedScale));
        secondaryProgressAnimator.addUpdateListener(secondaryProgressAnimationUpdateListener);
        secondaryProgressAnimator.addListener(secondaryProgressAnimationAdapterListener);
        secondaryProgressAnimator.start();
    }

    private ValueAnimator.AnimatorUpdateListener secondaryProgressAnimationUpdateListener = new ValueAnimator.AnimatorUpdateListener() {
        @Override
        public void onAnimationUpdate(ValueAnimator animation) {
            onUpdateSecondaryProgressByAnimation((float) animation.getAnimatedValue());
        }
    };

    private AnimatorListenerAdapter secondaryProgressAnimationAdapterListener = new AnimatorListenerAdapter() {
        @Override
        public void onAnimationEnd(Animator animation) {
            isSecondaryProgressAnimating = false;
            onSecondaryProgressAnimationEnd();
        }

        @Override
        public void onAnimationCancel(Animator animation) {
            isSecondaryProgressAnimating = false;
        }
    };

    private void onUpdateSecondaryProgressByAnimation(float progress) {
        super.setSecondaryProgress(progress);
        onProgressChangeAnimationUpdate(layoutSecondaryProgress, progress, lastProgress);
    }

    private void onSecondaryProgressAnimationEnd() {
        onProgressChangeAnimationEnd(layoutSecondaryProgress);
    }

    private void restoreSecondaryProgressAnimationState() {
        if (isSecondaryProgressAnimating) {
            startSecondaryProgressAnimation(super.getSecondaryProgress(), lastSecondaryProgress);
        }
    }

    private void clearSecondaryProgressAnimation() {
        if (secondaryProgressAnimator != null && secondaryProgressAnimator.isRunning()) {
            secondaryProgressAnimator.cancel();
        }
    }

    @Override
    protected Parcelable onSaveInstanceState() {
        Parcelable superState = super.onSaveInstanceState();
        SavedState ss = new SavedState(superState);

        ss.isProgressAnimating = this.isProgressAnimating;
        ss.isSecondaryProgressAnimating = this.isSecondaryProgressAnimating;
        ss.lastProgress = this.lastProgress;
        ss.lastSecondaryProgress = this.lastSecondaryProgress;
        ss.animationSpeedScale = this.animationSpeedScale;
        ss.isAnimationEnabled = this.isAnimationEnabled;

        return ss;
    }

    @Override
    protected void onRestoreInstanceState(Parcelable state) {
        if (!(state instanceof SavedState)) {
            super.onRestoreInstanceState(state);
            return;
        }

        SavedState ss = (SavedState) state;
        super.onRestoreInstanceState(ss.getSuperState());

        this.isProgressAnimating = ss.isProgressAnimating;
        this.isSecondaryProgressAnimating = ss.isSecondaryProgressAnimating;
        this.lastProgress = ss.lastProgress;
        this.lastSecondaryProgress = ss.lastSecondaryProgress;
        this.animationSpeedScale = ss.animationSpeedScale;
        this.isAnimationEnabled = ss.isAnimationEnabled;
        restoreProgressAnimationState();
        restoreSecondaryProgressAnimationState();
    }

    protected static class SavedState extends AbsSavedState {
        boolean isProgressAnimating;
        boolean isSecondaryProgressAnimating;
        float lastProgress;
        float lastSecondaryProgress;
        float animationSpeedScale;
        boolean isAnimationEnabled;

        SavedState(Parcelable superState) {
            super(superState);
        }

        SavedState(Parcel in) {
            this(in, null);
        }

        SavedState(Parcel in, ClassLoader loader) {
            super(in, loader);
            this.isProgressAnimating = in.readByte() != 0;
            this.isSecondaryProgressAnimating = in.readByte() != 0;
            this.lastProgress = in.readFloat();
            this.lastSecondaryProgress = in.readFloat();
            this.animationSpeedScale = in.readFloat();
            this.isAnimationEnabled = in.readByte() != 0;
        }

        @Override
        public void writeToParcel(Parcel out, int flags) {
            super.writeToParcel(out, flags);
            out.writeByte((byte) (this.isProgressAnimating ? 1 : 0));
            out.writeByte((byte) (this.isSecondaryProgressAnimating ? 1 : 0));
            out.writeFloat(this.lastProgress);
            out.writeFloat(this.lastSecondaryProgress);
            out.writeFloat(this.animationSpeedScale);
            out.writeByte((byte) (this.isAnimationEnabled ? 1 : 0));
        }

        public static final Parcelable.ClassLoaderCreator<SavedState> CREATOR = new Parcelable.ClassLoaderCreator<SavedState>() {
            @Override
            public SavedState createFromParcel(Parcel in, ClassLoader loader) {
                return new SavedState(in, loader);
            }

            public SavedState createFromParcel(Parcel in) {
                return new SavedState(in);
            }

            public SavedState[] newArray(int size) {
                return new SavedState[size];
            }
        };
    }
}

File: round-corner-progress-bar/src/main/java/com/akexorcist/roundcornerprogressbar/common/AnimatedRoundCornerProgressBar.kt
package com.akexorcist.roundcornerprogressbar.common

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.ValueAnimator
import android.animation.ValueAnimator.AnimatorUpdateListener
import android.content.Context
import android.os.Build
import android.os.Parcel
import android.os.Parcelable
import android.os.Parcelable.ClassLoaderCreator
import android.util.AttributeSet
import android.widget.LinearLayout
import androidx.annotation.FloatRange
import androidx.annotation.RequiresApi
import androidx.customview.view.AbsSavedState
import com.akexorcist.roundcornerprogressbar.R
import kotlin.math.abs

@Suppress("unused")
abstract class AnimatedRoundCornerProgressBar : BaseRoundCornerProgressBar {
    companion object {
        const val DEFAULT_DURATION = 500L
    }

    private var _isProgressAnimating = false
    private var _isSecondaryProgressAnimating = false
    private var _lastProgress = 0f
    private var _lastSecondaryProgress = 0f

    private var _animationSpeedScale = 1f
    private var _isAnimationEnabled = false

    private var _progressAnimator: ValueAnimator? = null
    private var _secondaryProgressAnimator: ValueAnimator? = null

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    )

    @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int, defStyleRes: Int) : super(
        context,
        attrs,
        defStyleAttr,
        defStyleRes
    )

    override fun setupStyleable(context: Context, attrs: AttributeSet?) {
        super.setupStyleable(context, attrs)
        if (attrs == null) return

        val typedArray = context.obtainStyledAttributes(
            attrs, R.styleable.AnimatedRoundCornerProgressBar
        )

        with(typedArray) {
            _isAnimationEnabled = getBoolean(
                R.styleable.AnimatedRoundCornerProgressBar_rcAnimationEnable, false
            )
            _animationSpeedScale = getFloat(
                R.styleable.AnimatedRoundCornerProgressBar_rcAnimationSpeedScale, 1f
            )
            recycle()
        }
        _lastProgress = super.getProgress()
        _lastSecondaryProgress = super.getSecondaryProgress()
    }

    override fun getProgress(): Float {
        return if (!_isAnimationEnabled && !_isProgressAnimating) {
            super.getProgress()
        } else {
            _lastProgress
        }
    }

    override fun setProgress(progress: Int) {
        setProgress(progress.toFloat())
    }

    override fun setProgress(progress: Float) {
        val actualProgress =
            if (progress < 0) 0f
            else progress.coerceAtMost(_max)

        clearProgressAnimation()
        this._lastProgress = actualProgress

        if (this._isAnimationEnabled) {
            startProgressAnimation(super.getProgress(), actualProgress)
        } else {
            super.setProgress(actualProgress)
        }
    }

    override fun getSecondaryProgress(): Float {
        return if (!_isAnimationEnabled && !_isSecondaryProgressAnimating) {
            super.getSecondaryProgress()
        } else {
            _lastSecondaryProgress
        }
    }

    override fun setSecondaryProgress(progress: Int) {
        setSecondaryProgress(progress.toFloat())
    }

    override fun setSecondaryProgress(progress: Float) {
        val actualProgress =
            if (progress < 0) 0f
            else progress.coerceAtMost(_max)
        clearSecondaryProgressAnimation()
        _lastSecondaryProgress = actualProgress
        if (_isAnimationEnabled) {
            startSecondaryProgressAnimation(super.getSecondaryProgress(), actualProgress)
        } else {
            super.setSecondaryProgress(actualProgress)
        }
    }

    @FloatRange(from = 0.2, to = 5.toDouble())
    open fun getAnimationSpeedScale(): Float {
        return _animationSpeedScale
    }

    fun enableAnimation() {
        _isAnimationEnabled = true
    }

    fun disableAnimation() {
        _isAnimationEnabled = false
    }

    fun setAnimationSpeedScale(@FloatRange(from = 0.2, to = 5.toDouble()) scale: Float) {
        _animationSpeedScale = scale.coerceIn(0.2f, 5f)
    }

    fun isProgressAnimating(): Boolean = _isProgressAnimating

    fun isSecondaryProgressAnimating(): Boolean = _isSecondaryProgressAnimating

    protected open fun onProgressChangeAnimationUpdate(
        layout: LinearLayout,
        current: Float,
        to: Float
    ) {
    }

    protected open fun onProgressChangeAnimationEnd(layout: LinearLayout) {}

    protected open fun stopProgressAnimationImmediately() {
        clearProgressAnimation()
        clearSecondaryProgressAnimation()
        if (_isAnimationEnabled && _isProgressAnimating) {
            disableAnimation()
            if (_isProgressAnimating) {
                super.setProgress(_lastProgress)
            }
            if (_isSecondaryProgressAnimating) {
                super.setSecondaryProgress(_lastProgress)
            }
            enableAnimation()
        }
    }

    private fun getAnimationDuration(from: Float, to: Float, max: Float, scale: Float): Long {
        val diff = abs(from - to)
        return (diff * DEFAULT_DURATION / max * scale).toLong()
    }

    private fun startProgressAnimation(from: Float, to: Float) {
        _isProgressAnimating = true
        _progressAnimator?.apply {
            removeUpdateListener(progressAnimationUpdateListener)
            removeListener(progressAnimationAdapterListener)
            cancel()
        }
        _progressAnimator = ValueAnimator.ofFloat(from, to).apply {
            duration = getAnimationDuration(from, to, _max, _animationSpeedScale)
            addUpdateListener(progressAnimationUpdateListener)
            addListener(progressAnimationAdapterListener)
            start()
        }
    }

    private val progressAnimationUpdateListener = AnimatorUpdateListener { animation ->
        onUpdateProgressByAnimation(animation.animatedValue as Float)
    }

    private val progressAnimationAdapterListener: AnimatorListenerAdapter =
        object : AnimatorListenerAdapter() {
            override fun onAnimationEnd(animation: Animator) {
                _isProgressAnimating = false
                onProgressAnimationEnd()
            }

            override fun onAnimationCancel(animation: Animator) {
                _isProgressAnimating = false
            }
        }

    private fun onUpdateProgressByAnimation(progress: Float) {
        super.setProgress(progress)
        onProgressChangeAnimationUpdate(layoutProgress, progress, _lastProgress)
    }

    private fun onProgressAnimationEnd() {
        onProgressChangeAnimationEnd(layoutProgress)
    }

    private fun restoreProgressAnimationState() {
        if (_isProgressAnimating) {
            startProgressAnimation(super.getProgress(), _lastProgress)
        }
    }

    private fun clearProgressAnimation() {
        if (_progressAnimator?.isRunning == true) {
            _progressAnimator?.cancel()
        }
    }

    private fun startSecondaryProgressAnimation(from: Float, to: Float) {
        _isSecondaryProgressAnimating = true
        _secondaryProgressAnimator?.apply {
            removeUpdateListener(secondaryProgressAnimationUpdateListener)
            removeListener(secondaryProgressAnimationAdapterListener)
            cancel()
        }
        _secondaryProgressAnimator = ValueAnimator.ofFloat(from, to).apply {
            duration = getAnimationDuration(from, to, _max, _animationSpeedScale)
            addUpdateListener(secondaryProgressAnimationUpdateListener)
            addListener(secondaryProgressAnimationAdapterListener)
            start()
        }
    }

    private val secondaryProgressAnimationUpdateListener = AnimatorUpdateListener { animation ->
        onUpdateSecondaryProgressByAnimation(animation.animatedValue as Float)
    }

    private val secondaryProgressAnimationAdapterListener: AnimatorListenerAdapter =
        object : AnimatorListenerAdapter() {
            override fun onAnimationEnd(animation: Animator) {
                _isSecondaryProgressAnimating = false
                onSecondaryProgressAnimationEnd()
            }

            override fun onAnimationCancel(animation: Animator) {
                _isSecondaryProgressAnimating = false
            }
        }

    private fun onUpdateSecondaryProgressByAnimation(progress: Float) {
        super.setSecondaryProgress(progress)
        onProgressChangeAnimationUpdate(layoutSecondaryProgress, progress, _lastProgress)
    }

    private fun onSecondaryProgressAnimationEnd() {
        onProgressChangeAnimationEnd(layoutSecondaryProgress)
    }

    private fun restoreSecondaryProgressAnimationState() {
        if (_isSecondaryProgressAnimating) {
            startSecondaryProgressAnimation(super.getSecondaryProgress(), _lastSecondaryProgress)
        }
    }

    private fun clearSecondaryProgressAnimation() {
        if (_secondaryProgressAnimator?.isRunning == true) {
            _secondaryProgressAnimator?.cancel()
        }
    }

    override fun onSaveInstanceState(): Parcelable? {
        val superState = super.onSaveInstanceState() ?: return null
        val state = SavedState(superState)
        with(state) {
            isProgressAnimating = _isProgressAnimating
            isSecondaryProgressAnimating = _isSecondaryProgressAnimating
            lastProgress = _lastProgress
            lastSecondaryProgress = _lastSecondaryProgress
            animationSpeedScale = _animationSpeedScale
            isAnimationEnabled = _isAnimationEnabled
        }
        return state
    }

    override fun onRestoreInstanceState(state: Parcelable?) {
        if (state !is SavedState) {
            super.onRestoreInstanceState(state)
            return
        }
        super.onRestoreInstanceState(state.superState)
        with(state) {
            _isProgressAnimating = isProgressAnimating
            _isSecondaryProgressAnimating = isSecondaryProgressAnimating
            _lastProgress = lastProgress
            _lastSecondaryProgress = lastSecondaryProgress
            _animationSpeedScale = animationSpeedScale
            _isAnimationEnabled = isAnimationEnabled
        }
        restoreProgressAnimationState()
        restoreSecondaryProgressAnimationState()
    }

    protected class SavedState : AbsSavedState {
        var isProgressAnimating = false
        var isSecondaryProgressAnimating = false
        var lastProgress = 0f
        var lastSecondaryProgress = 0f
        var animationSpeedScale = 0f
        var isAnimationEnabled = false

        constructor(superState: Parcelable) : super(superState)

        constructor(source: Parcel) : super(source)

        constructor(source: Parcel, loader: ClassLoader?) : super(source, loader) {
            with(source) {
                isProgressAnimating = readByte().toInt() != 0
                isSecondaryProgressAnimating = readByte().toInt() != 0
                lastProgress = readFloat()
                lastSecondaryProgress = readFloat()
                animationSpeedScale = readFloat()
                isAnimationEnabled = readByte().toInt() != 0
            }
        }

        override fun writeToParcel(dest: Parcel, flags: Int) {
            super.writeToParcel(dest, flags)
            with(dest) {
                writeByte((if (isProgressAnimating) 1 else 0).toByte())
                writeByte((if (isSecondaryProgressAnimating) 1 else 0).toByte())
                writeFloat(lastProgress)
                writeFloat(lastSecondaryProgress)
                writeFloat(animationSpeedScale)
                writeByte((if (isAnimationEnabled) 1 else 0).toByte())
            }
        }

        companion object {
            @JvmField
            val CREATOR: ClassLoaderCreator<SavedState> = object : ClassLoaderCreator<SavedState> {
                override fun createFromParcel(source: Parcel, loader: ClassLoader): SavedState =
                    SavedState(source, loader)

                override fun createFromParcel(source: Parcel): SavedState = SavedState(source)

                override fun newArray(size: Int): Array<SavedState> = newArray(size)
            }
        }
    }
}
